### PR TITLE
Fix update of plugins and disable update button if not available on conda forge (for bundle)

### LINKED ISSUE #4427: Plugin update in bundle does not work
## 🐛 Bug

To update the plugin I need to uninstall it and then install it again. Clicking update trigger a processing event but ends without a result

Zrzut ekranu z 2022-04-26 18-44-25

## To Reproduce

1. install bundled (new with conda) napari
2. install old version plugin
3. try to update it from GUI 

## Expected behavior

Update should work

## Environment
```
napari: 0.4.15
Platform: Linux-5.13.0-40-generic-x86_64-with-glibc2.34
System: Ubuntu 21.10
Python: 3.9.10 | packaged by conda-forge | (main, Feb 1 2022, 21:24:11) [GCC 9.4.0]
Qt: 5.12.9
PySide2: 5.13.2
NumPy: 1.22.3
SciPy: 1.8.0
Dask: 2022.02.1
VisPy: 0.9.6

OpenGL:
- GL version: 4.6.0 NVIDIA 510.60.02
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0
- screen 2: resolution 1920x1080, scale 1.0
- screen 3: resolution 1920x1080, scale 1.0

Plugins:
- PartSeg Analysis project reader: 0.14.0
- PartSeg Image reader: 0.14.0
- PartSeg Image tiff writer: 0.14.0
- PartSeg Mask project reader: 0.14.0
- PartSeg Mask project writer: 0.14.0
- PartSeg Masked Tiff reader: 0.14.0
- PartSeg widgets: 0.14.0
- console: 0.0.4
- scikit-image: 0.4.15
- svg: 0.1.5
```

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #4427

# Description

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [x] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
@Czaki this should fix the issue you found.

File: napari/_qt/dialogs/qt_plugin_dialog.py [hunk 1]
```diff
@@ -656,7 +654,10 @@ def handle_action(self, item, pkg_name, action_name, update=False):
             method((pkg_name,))
 
     @Slot(PackageMetadata)
-    def tag_outdated(self, project_info: PackageMetadata):
+    def tag_outdated(self, project_info: PackageMetadata, is_available: bool):
```
Comment (Czaki) [inline napari/_qt/dialogs/qt_plugin_dialog.py]: Slot does not fit to function signature

File: napari/_qt/dialogs/qt_plugin_dialog.py [hunk 2]
```diff
@@ -998,7 +1000,7 @@ def _install_packages(self, packages: Sequence[str] = ()):
     def _handle_yield(self, data: Tuple[PackageMetadata, bool]):
         project_info, is_available = data
         if project_info.name in self.already_installed:
-            self.installed_list.tag_outdated(project_info)
+            self.installed_list.tag_outdated(project_info, is_available)
```
Comment (Czaki) [inline napari/_qt/dialogs/qt_plugin_dialog.py:1003]: As I inspect `hub_plugin_info` function from `napari.plugins.hub` its only check if most recent version is available on conda-forge. 

But if the user has a package in version `0.1` and versions `0.2` and `0.3` are published. But `0.3` does not reach yet conda -forge then he will not see a button with information that an update is available. But should. 

This also does not solve the problems of the package that the most recent version requires never dependency than it is pinned as core napari dependecy.

**Review by Czaki — CHANGES_REQUESTED** [review]

File: napari/_qt/dialogs/qt_plugin_dialog.py (hunk 1 above)
Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: You are right!, will fix it

File: napari/_qt/dialogs/qt_plugin_dialog.py (hunk 2 above)
Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py:1003]: >But if the user has a package in version 0.1 and versions 0.2 and 0.3 are published. But 0.3 does not reach yet conda -forge then he will not see a button with information that an update is available. But should.

Yes that is correct, but to solve this we need more information from the hub (which currently only lists the latest from pypi). So in the meantime is the best solution we can offer.

Pinging @neuromusic 

>This also does not solve the problems of the package that the most recent version requires never dependency than it is pinned as core napari dependecy.

I do not understand what you mean. Could you elaborate? thanks!

Comment (Czaki) [reply napari/_qt/dialogs/qt_plugin_dialog.py:1003]: > I do not understand what you mean. Could you elaborate? thanks!

Simple example. Napari bundle pins `superqt` to `0.3.2` In the meantime superqt `0.3.3` is released. Someone decides that he needs a feature from `0.3.3` and sets it as a minimum dependency. Create release. upload to pypi and anaconda. 

The current code will report that there is an update, but every installation will fail without explanation. 

> Yes that is correct, but to solve this we need more information from the hub (which currently only lists the latest from pypi). So in the meantime is the best solution we can offer.

No, you depend also on information from conda so you could calculate maximum available version from here: 

https://github.com/napari/napari/blob/3b3e2d1d9c63da141144e500b838ee9ea4b585cb/napari/plugins/hub.py#L56-L59

Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py:1003]: >The current code will report that there is an update, but every installation will fail without explanation.

Yes but that is a separate issue, please open one, or refer to the one where we are discussing the metapackage pinnings for napari. Also that issue is out of the scope of the initial issue described and this PR.

Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py:1003]: >No, you depend also on information from conda so you could calculate maximum available version from here:

You are right, added a fix!

File: napari/plugins/hub.py [hunk 3]
```diff
@@ -55,9 +55,10 @@ def hub_plugin_info(
         try:
             with request.urlopen(anaconda_api) as resp_api:
                 anaconda_info = json.loads(resp_api.read().decode())
-                if version in anaconda_info.get("versions", []):
-                    is_available_in_conda_forge = True
-
+                versions = anaconda_info.get("versions", [])
+                is_available_in_conda_forge = True
+                if version not in versions:
+                    version = versions[-1]
```
Comment (Czaki) [inline napari/plugins/hub.py]:
```suggestion
                versions = anaconda_info.get("versions", [])
                is_available_in_conda_forge = bool(versions)
                if is_available_in_conda_forge and version not in versions:
                    version = versions[-1]
```

Are you sure that anaconda API return versions in sorted order? Or sort steep is missed?

Comment (goanpeca) [reply napari/plugins/hub.py]: >Are you sure that anaconda API return versions in sorted order? Or sort steep is missed?

Yes they are returned in sorted order

Comment (Czaki) [reply napari/plugins/hub.py]: Nice, but still, the code needs to handle an empty list properly.

Comment (goanpeca) [reply napari/plugins/hub.py]: Fixed and added tests @Czaki.

**Review by Czaki — APPROVED** [review]
Thanks

**Review by jni — APPROVED** [review]
Thanks for the fix @goanpeca and for the thorough review @Czaki! 🎉

File: napari/_qt/dialogs/qt_plugin_dialog.py (hunk 2 above)
Comment (jaimergp) [reply napari/_qt/dialogs/qt_plugin_dialog.py:1003]: > Napari bundle pins superqt to 0.3.2

We do not freeze it (yet); we only freeze `napari` itself, not its immediate dependencies. So it _should_ work, but we need to control the pinnings to mark _how_ we are going to allow this to happen (for example, superqt can be updated as long as is version 0.3.x).

**Pull request merged**